File: packages/node/shared-infra/src/http/resilience/concurrency-admission-control.ts
/**
 * @file concurrency-admission-control.ts
 * @description Inbound Concurrency Load Shedding & Capacity Manager.
 * 
 * ALGORITHM & SPECIFICATION:
 * 1. Process-Level Capacity Guard:
 *    - Tracks active in-flight requests via atomic counter (`activeInFlight`).
 *    - Enforces maximum concurrent in-flight limit (`maxCapacity`, default: 500).
 * 2. Admission Decision Mechanics:
 *    - `acquire()`: If `activeInFlight < maxCapacity`, increments counter and returns `true` (Admitted).
 *    - If `activeInFlight >= maxCapacity`, returns `false` (Shed/Rejected).
 *    - `release()`: Decrements `activeInFlight` safely (bounded to 0).
 */

export class ConcurrencyAdmissionControl {
  private activeInFlight = 0;

  constructor(private readonly maxCapacity = 500) {}

  public acquire(): boolean {
    if (this.activeInFlight >= this.maxCapacity) {
      return false;
    }
    this.activeInFlight++;
    return true;
  }

  public release(): void {
    if (this.activeInFlight > 0) {
      this.activeInFlight--;
    }
  }

  public getActiveCount(): number {
    return this.activeInFlight;
  }

  public getMaxCapacity(): number {
    return this.maxCapacity;
  }
}

File: packages/llm-obs-python-packages/python-shared/src/python_shared/http/resilience/concurrency_admission_control.py
import threading

class ConcurrencyAdmissionControl:
    def __init__(self, max_capacity: int = 500):
        self.max_capacity = max_capacity
        self.active_in_flight = 0
        self._lock = threading.RLock()

    def acquire(self) -> bool:
        with self._lock:
            if self.active_in_flight >= self.max_capacity:
                return False
            self.active_in_flight += 1
            return True

    def release(self) -> None:
        with self._lock:
            if self.active_in_flight > 0:
                self.active_in_flight -= 1

    def get_active_count(self) -> int:
        with self._lock:
            return self.active_in_flight

    def get_max_capacity(self) -> int:
        return self.max_capacity
